### PR TITLE
Add mocked cores and nodes to ActionTesting

### PR DESCRIPTION
**Pull request by markscheel**
## Proposed changes

The ActionTesting framework now can simulate running on multiple nodes and cores.

MockRuntimeSystem now has a `std::vector<size_t>` passed into its constructor, specifying the number of (mock) cores on each (mock) node of the mocked runtime system.  Each `MockDistributedObject` and their proxies keep track of the mock node and core that it lives on. The various member functions in ActionTesting have been updated to do the right thing for multiple nodes and cores.

Test_ActionTesting tests much of the new functionality.

Many of the other tests have been updated to work with the new changes.

### Upgrade instructions

- In ActionTesting, if you are mocking a Group, Nodegroup, or Singleton component, these should now have `chare_type` set to `MockGroupChare`, `MockNodeGroupChare`, or `MockSingletonChare` instead of `MockArrayChare`.  Note that `MockArrayChare` will *sometimes* work for these components for some tests, but not if `ckLocal` or `ckLocalBranch` is ever called on those components.
- In ActionTesting, the new functions `emplace_singleton_component`, `emplace_group_component`, `emplace_array_component`, and `emplace_nodegroup_component` should be used instead of `emplace_component`.  Note that `emplace_component` still works but will be removed in the future.
- The functions in Utilities/System/ParallelInfo now have two versions: the low-level charm++ version (in the `sys` namespace) and a new version templated on the `DistributedObject`.  The `sys` version should be used only for low-level things like `AbortWithErrorMessage` or the `ParallelInfo` executable where you know they will never be mocked; the templated version should be used everywhere else.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Parallel/Invoke.hpp
Comment (wthrowe) [inline src/Parallel/Invoke.hpp]: `ckLocal_disabled` and again two lines down.

File: src/Parallel/Invoke.hpp
Comment (wthrowe) [inline src/Parallel/Invoke.hpp]: Why not `ckLocal_enabled` to match the thing being implemented and avoid inverted logic?

File: tests/Unit/Framework/MockDistributedObject.hpp
Comment (wthrowe) [inline tests/Unit/Framework/MockDistributedObject.hpp]: Take vectors by value and move.

File: tests/Unit/Framework/ActionTesting.hpp
Comment (wthrowe) [inline tests/Unit/Framework/ActionTesting.hpp:72]: Commit message: (ActionTesting: Add mocked nodes and cores.)  "There are new functions..." should emplace_array_component be included?

File: tests/Unit/Framework/ActionTesting.hpp
Comment (wthrowe) [inline tests/Unit/Framework/ActionTesting.hpp:616]: Why is this calling the action on every component?

File: tests/Unit/Framework/Tests/Test_ActionTesting.cpp
Comment (wthrowe) [inline tests/Unit/Framework/Tests/Test_ActionTesting.cpp]: Remove line.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Parallel/Invoke.hpp
Comment (markscheel) [reply src/Parallel/Invoke.hpp]: Sure, I can call it `ckLocal_enabled`. Are you ok with structs that don't define `ckLocal_enabled` to be considered enabled?

File: tests/Unit/Framework/ActionTesting.hpp
Comment (markscheel) [reply tests/Unit/Framework/ActionTesting.hpp:616]: Because this is a member function of MockCollectionOfDistributedObjectsProxy, i.e. a proxy to the entire chare, not to one element of the chare.  So just like simple_action, shouldn't threaded_action be a broadcast call to all elements?  If you want one element, you would use ckLocalBranch to get your local element (or index the MockCollectionOfDistributedObjectsProxy to get a different element if you want that), and that would return a MockDistributedObject (in the case of ckLocalBranch) or a proxy to one (in the case of indexing) that you could call threaded_action on for a single element.

Pls let me know if my understanding of threaded_action is incorrect.

File: src/Parallel/Invoke.hpp
Comment (wthrowe) [reply src/Parallel/Invoke.hpp]: Yes.

File: tests/Unit/Framework/ActionTesting.hpp
Comment (wthrowe) [reply tests/Unit/Framework/ActionTesting.hpp:616]: You can't run a threaded action on an entire component, and you can't even run it on a remote element.  I thought you couldn't even run it on a proxy, but our documentation and code disagree with each other on this:

docs/DevGuide/Parallelization.md:
> User controlled threading is possible by calling the non-entry method member function `threaded_action`.

src/Parallel/Algorithms/AlgorithmNodegroup.ci:
> template \<typename Action>
> entry void threaded_action();

@nilsdeppe which of these is correct?

Comment (markscheel) [reply tests/Unit/Framework/ActionTesting.hpp:616]: @wthrowe If we aren't supposed to ever run a threaded action on an entire component, then `MockCollectionOfDistributedObjectsProxy::threaded_action` should just be removed; so that's what I did in a fixup commit. All the tests pass, so we never used that function anyway.

**Comment by markscheel** [comment]
Fixups to address review by @wthrowe

File: tests/Unit/Framework/MockDistributedObject.hpp
Comment (moxcodes) [inline tests/Unit/Framework/MockDistributedObject.hpp]: add backticks to the code parts of the documentation.

File: tests/Unit/Framework/ActionTesting.hpp
Comment (nilsdeppe) [reply tests/Unit/Framework/ActionTesting.hpp:616]: The way we implement threaded entry methods they can be run on an entire component. I don't think we use that feature right now. I see no reason right now to disable it either. The threaded actions should behave (almost) entirely the same way as simple actions except that they don't lock the entire group.

**Comment by markscheel** [comment]
Restored threaded_action.

File: tests/Unit/Framework/MockDistributedObject.hpp
Comment (wthrowe) [inline tests/Unit/Framework/MockDistributedObject.hpp]: No const.

**Review by wthrowe — CHANGES_REQUESTED** [review]
Squash directly.

File: tests/Unit/Framework/MockDistributedObject.hpp
Comment (markscheel) [reply tests/Unit/Framework/MockDistributedObject.hpp]: oops. Yes, of course.

**Comment by markscheel** [comment]
Fixed MockDistributedObject and squashed.

**Review by wthrowe — DISMISSED** [review]

File: tests/Unit/Framework/ActionTesting.hpp
Comment (moxcodes) [inline tests/Unit/Framework/ActionTesting.hpp:696]: if it is the same for all possible specializations, why specify four times? I suppose this way it causes a compile-time error if something tries to use it on a non-mock-chare type, but you'd get a compile-time error when it tried to access the `::array_index` anyways.

File: tests/Unit/Framework/MockRuntimeSystem.hpp
Comment (moxcodes) [inline tests/Unit/Framework/MockRuntimeSystem.hpp]: [optional] This inner loop can be simplified with `std::iota`. I felt like the outer loop could also be replaced with a sufficiently clever `std::inclusive_scan` but it wasn't actually shorter than the explicit loop.

File: tests/Unit/Framework/MockRuntimeSystem.hpp
Comment (moxcodes) [inline tests/Unit/Framework/MockRuntimeSystem.hpp:231]: Do I understand correctly that these instances of using `[array_index]` with the `inboxes_` works primarily because its an unordered map, and therefore this line is doing the work of both inserting the new element, then passing a pointer to that new element to the new `MockDistributedObject`?
That's a little confusing (and a reader is likely going to look around for where `inboxes_` got initialized and fail to find anything). Perhaps either a code comment explaining the trick, or use a `.insert` on a line before the `emplace` to separate out the two operations. 
(any edits should apply to the 8 similar `emplace` functions)

File: tests/Unit/Framework/Tests/Test_ActionTesting.cpp
Comment (moxcodes) [inline tests/Unit/Framework/Tests/Test_ActionTesting.cpp]: [nitpick] has this been formatted? I would have expected more space in the initializer list.

File: src/Utilities/System/ParallelInfo.hpp
Comment (moxcodes) [inline src/Utilities/System/ParallelInfo.hpp]: I'm confused... don't functions elsewhere use the functions in this namespace? Why didn't other things have to change?

File: src/Elliptic/DiscontinuousGalerkin/DgElementArray.hpp
Comment (moxcodes) [inline src/Elliptic/DiscontinuousGalerkin/DgElementArray.hpp]: Oh, I see..  This commit should be combined with the one that's the namespace change so everything still compiles on each commit.

**Review by moxcodes — COMMENTED** [review]
I just have a few minor comments -- feel free to squash the changes to these.

**Comment by markscheel** [comment]
Made changes suggested by @moxcodes including squashing the sys::charm namespace commits.  Squashed.

**Comment by moxcodes** [comment]
Did something with git go wrong? I think the commits are still the same, including the separate namespace commits.

**Comment by markscheel** [comment]
Yeah, something with git went wrong.  Need to redo because my local copy doesn't look like it should.

File: tests/Unit/Framework/ActionTesting.hpp
Comment (markscheel) [reply tests/Unit/Framework/ActionTesting.hpp:696]: Thinking about this more: Yes, I could avoid 4 specializations, but I don't see why you would get a compile-time error when trying to access `array_index`.  Couldn't someone specify in a Component a misspelled `using chare_type=MockArrayyyChare` and a correct `using array_index=size_t`, and then accessing array_index would work but there may be a more confusing compiler message elsewhere?

Comment (moxcodes) [reply tests/Unit/Framework/ActionTesting.hpp:696]: Yeah, that's a reasonable point -- this probably does avoid some misbehavior earlier than it would if were just a generic template. I'm fine with it being left as-is.

**Comment by markscheel** [comment]
Made and squashed @moxcodes suggestions.  Also added a new fixup at request of @nilsleiffischer who wanted the node/core functions in Parallel/Info.hpp

File: src/Parallel/Info.hpp
Comment (nilsvu) [inline src/Parallel/Info.hpp]: Could you make this the `Parallel` namespace and the one in `Utilities/System` just `sys` (instead of `sys::charm`)? Then if I understand correctly this PR keeps the low-level `sys` functions the same but introduces new `Parallel` functions that work with distributed objects.

Comment (markscheel) [reply src/Parallel/Info.hpp]: I can put these into the `Parallel` namespace.  But I kind of like `sys::charm` for the other ones to emphasize that these are charm-dependent functions, and also to discourage people from using them (or make people think twice before using them).

One worry I have:  Suppose someone writes a new Action (or modifies an existing Action) and uses the charm functions. There will be no compiler error; the only effect will be that during ActionTesting the function will return the wrong answer (i.e. always one core and one node) and result in a bug that will be hard to find.  Thoughts?

Comment (nilsvu) [reply src/Parallel/Info.hpp]: - Yea we can consider moving the "raw" functions from `sys` -> `sys::charm`, though I think that doesn't need to happen in this PR. I would prefer to keep them in `sys` because they are not inherently charm-specific (only their implementation calls into charm).
- I can't think of a foolproof way to make sure people use the `Parallel` functions, other than adding warnings to the docs. The `sys` functions have legitimate use-cases in layers of code that don't yet depend on `Parallel` (mostly error-handling) to avoid cyclic dependencies.
- The interface you suggested before, where the distributed-object functions are in `sys`, has the problem that distributed objects are a `Parallel` concept that shouldn't be lifted to `sys`. Since the functions are generic the cyclic dependency is technically broken (I think). However, the only argument I can see for putting the distributed-object functions in `sys` is to "hide" the raw functions a bit more by moving the `Parallel` functions closer to them, which doesn't really address the worry that people can use the raw functions.

So my suggestion is: Keep the raw functions in `sys`. Add functions to `Parallel` that work with distributed objects. Add warnings to the `sys` functions that they are super low-level and people should use the `Parallel` functions when they work with distributed objects. What do you think?

Comment (markscheel) [reply src/Parallel/Info.hpp]: Ok, done!  Fixup commits pushed (and rebase onto current develop)

**Review by moxcodes — DISMISSED** [review]

File: src/Parallel/Info.hpp
Comment (wthrowe) [inline src/Parallel/Info.hpp]: Consider making this Doxygen.

File: src/ParallelAlgorithms/LinearSolver/Observe.hpp
Comment (wthrowe) [inline src/ParallelAlgorithms/LinearSolver/Observe.hpp]: Alphabet.

**Review by wthrowe — COMMENTED** [review]
I'm happy with the fixups.  My requests can be squashed directly.

File: src/Parallel/Info.hpp
Comment (nilsvu) [reply src/Parallel/Info.hpp]: also there's a residual `sys::charm` -> `sys` in line 13

**Comment by nilsvu** [comment]
My concern in the thread https://github.com/sxs-collaboration/spectre/pull/2792#discussion_r568473959 is addressed. I think there's no need I jump in as another full reviewer, so no need to wait for my approval.

**Comment by markscheel** [comment]
Fixed the remaining things, squashed everything, rebased on develop.

**Review by moxcodes — APPROVED** [review]

File: tests/Unit/Framework/MockDistributedObject.hpp
Comment (kidder) [inline tests/Unit/Framework/MockDistributedObject.hpp:727]: how about `std::unordered_map<NodeId, std::unordered_map<LocalCoreId, GlobalCoreId>>`

File: tests/Unit/Framework/MockDistributedObject.hpp
Comment (kidder) [inline tests/Unit/Framework/MockDistributedObject.hpp:729]: and here `std::unordered_map<GlobalCoreId, std::pair<NodeId,LocalCoreId>>`

File: tests/Unit/Framework/ActionTesting.hpp
Comment (kidder) [inline tests/Unit/Framework/ActionTesting.hpp:580]: template this on   `template <typename U = ChareType, Requires<std::is_same_v<U, ChareType> and std::is_same_v<U, MockSingletonChare>>>` and then you shouldn't need `ckLocal_enabled`.

File: tests/Unit/Framework/ActionTesting.hpp
Comment (kidder) [inline tests/Unit/Framework/ActionTesting.hpp:646]: typo: of the

File: tests/Unit/Framework/MockRuntimeSystem.hpp
Comment (kidder) [inline tests/Unit/Framework/MockRuntimeSystem.hpp:174]: on each core?

File: tests/Unit/Framework/MockRuntimeSystem.hpp
Comment (kidder) [inline tests/Unit/Framework/MockRuntimeSystem.hpp:326]: use a structured binding [iterator, a better_name_than_bool].   (Repeat below)

File: tests/Unit/Framework/MockRuntimeSystem.hpp
Comment (kidder) [inline tests/Unit/Framework/MockRuntimeSystem.hpp:393]: structured binding [node, global_core]

File: tests/Unit/Framework/MockRuntimeSystem.hpp
Comment (kidder) [inline tests/Unit/Framework/MockRuntimeSystem.hpp:704]: replace mock_global_cores_  and mock_nodes_and_local_cores_ with same types as commented on in MockDistributedObject

**Review by wthrowe — APPROVED** [review]

File: src/Parallel/Info.hpp
Comment (kidder) [inline src/Parallel/Info.hpp:24]: noexcept everything

**Review by kidder — COMMENTED** [review]
Reviewed with @nilsdeppe.   Please consider a followup PR to address these minor changes that we think will improve usability

**Pull request merged**
